packages/flutter_tools/lib/src/platform_step.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'android/apk.dart';
+import 'base/context.dart';
+import 'build_info.dart';
+import 'cache.dart';
+import 'project.dart';
+
+/// The [PlatformBuilders] instance.
+PlatformBuilders get platformBuilders => context.get<PlatformBuilders>();
+
+/// A registry that selects the correct [PlatformBuilder] based on the provided
+/// build info.
+class PlatformBuilders {
+  /// Create a new [PlatformBuilder] with `_platformSteps` as platform specific
+  /// implementations.
+  const PlatformBuilders([this._platformSteps = const <PlatformBuildStep>[
+    AndroidPlatformBuildStep(),
+  ]]);
+
+  final List<PlatformBuildStep> _platformSteps;
+
+  /// Build the platform specific bundle for [flutterProject].
+  ///
+  /// Selects the [PlatformBuildStep] that supports the required [TargetPlatform]
+  /// and [BuildMode] requested.
+  ///
+  /// Throws an [AmbiguousPlatformBuilder] if multiple build steps
+  /// match.
+  ///
+  /// Throws a [MissingPlatformBuilder] if no build steps match.
+  PlatformBuildStep selectPlatform({
+    BuildInfo buildInfo,
+  }) {
+    PlatformBuildStep selected;
+    for (PlatformBuildStep platformStep in _platformSteps) {
+      final Set<BuildMode> modes = platformStep.supported[buildInfo.targetPlatform];
+      if (modes == null) {
+        continue;
+      }
+      if (!modes.contains(buildInfo.mode)) {
+        continue;
+      }
+      if (selected != null) {
+        throw AmbiguousPlatformBuilder(
+          buildInfo.mode,
+          buildInfo.targetPlatform,
+        );
+      }
+      selected = platformStep;
+    }
+    if (selected == null) {
+      throw MissingPlatformBuilder(
+        buildInfo.mode,
+        buildInfo.targetPlatform,
+      );
+    }
+    return selected;
+  }
+}
+
+/// An exception thrown when there is no builder corresponding to the provided
+/// target platform and build mode combination.
+class MissingPlatformBuilder implements Exception {
```

Comment:
Maybe subclass, like "PlatformBuilderException" to make it easier to catch just the right things.